File: BOJ/14502_연구소.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#define X first
#define Y second
#define endl '\n'
using namespace std;

int n, m;
int result = -1;
int dx[4] = { 1, -1, 0, 0 };
int dy[4] = { 0, 0, 1, -1 };
vector<vector<int>> map;
vector<vector<int>> tmp;
vector<vector<bool>> visited;
vector<pair<int, int>> virus;
vector<pair<int, int>> zero;
pair<int, int> wall[3];
queue<pair<int, int>> q;

void bfs() {
	int curRes = 0;
	visited = vector<vector<bool>>(n, vector<bool>(m, false));
	tmp = map;

	for (int i = 0; i < 3; i++) {
		tmp[wall[i].X][wall[i].Y] = 1;
	}

	for (int i = 0; i < virus.size(); i++) {
		q.push(virus[i]);
		visited[virus[i].X][virus[i].Y] = true;
	}

	while (!q.empty()) {
		for (int i = 0; i < 4; i++) {
			int nx = dx[i] + q.front().X, ny = dy[i] + q.front().Y;
			
			if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
			if (!visited[nx][ny] && tmp[nx][ny] == 0) {
				visited[nx][ny] = true;
				tmp[nx][ny] = 2;
				q.push({ nx, ny });
			}
		}
		q.pop();
	}


	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			if (tmp[i][j] == 0) curRes++;
		}
	}
	result = max(result, curRes);
	return;
}

int main () {
	cin >> n >> m;
	map = vector<vector<int>>(n, vector<int>(m));

File: BOJ/14502_연구소.java
import java.io.*;
import java.util.*;

class Loc {
	int x,y;
	Loc(int x, int y ) {
		this.x = x;
		this.y = y;
	}
}

public class Main {
	static final int[] dx = {-1, 1, 0, 0};
	static final int[] dy = {0, 0, -1, 1};
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer tk;
	static int n, m;
	static int[][] map;
	static ArrayList<Loc> emp = new ArrayList<>();
	
	static int bfs(Loc a, Loc b, Loc c) {
		Queue<Loc> q = new LinkedList<>();
		int[][] tmp = new int[n][m];
		boolean[][] visited = new boolean[n][m];

		for(int i=0; i<n; i++) {
			for(int j=0; j<m; j++) {
				tmp[i][j] = map[i][j];
				if(tmp[i][j] == 2) {
					q.add(new Loc(i, j));
					visited[i][j] = true;
				}
			}
		}
		tmp[a.x][a.y] = tmp[b.x][b.y] = tmp[c.x][c.y] = 1;
		
		while(!q.isEmpty()) {
			Loc cur = q.poll();
			for(int i=0; i<4; i++) {
				int nx = cur.x + dx[i], ny = cur.y + dy[i];
				if(nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny] && tmp[nx][ny] == 0) {
					visited[nx][ny] = true;
					tmp[nx][ny] = 2;
					q.add(new Loc(nx, ny));
				}
			}
		}
		
		int cnt = 0;
		for(int i=0; i<n; i++) {
			for(int j=0; j<m; j++) {
				if(tmp[i][j] == 0) cnt++;
			}
		}
		return cnt;
	}
	
	public static void main(String[] args) throws Exception {
		tk = new StringTokenizer(br.readLine());
		n = Integer.parseInt(tk.nextToken());
		m = Integer.parseInt(tk.nextToken());
		
		map = new int[n][m];
		for(int i=0; i<n; i++) {
			tk = new StringTokenizer(br.readLine());
			for(int j=0; j<m; j++) {
				map[i][j] = Integer.parseInt(tk.nextToken());
				if(map[i][j] == 0 ) emp.add(new Loc(i, j));
			}
		}
		int result = Integer.MIN_VALUE;
		for(int i=0; i<emp.size()-2; i++) {
			for(int j=i+1; j<emp.size()-1; j++) {
				for(int k=j+1; k<emp.size(); k++) {
					Loc a = emp.get(i), b = emp.get(j), c = emp.get(k);
					if(map[a.x][a.y] == 0 && map[b.x][b.y] == 0 && map[c.x][c.y] == 0)
						result = Math.max(result, bfs(emp.get(i), emp.get(j), emp.get(k)));
				}
			}
		}
		System.out.println(result);
		
	}

}

